File: skills/tokensmind-agent-network/scripts/memory_actions.py
import math
from urllib.parse import quote, urlencode

from action_support import ActionState
from action_validation import require_text, text

MEMORY_KINDS = frozenset(("agent_experience", "matching_preference"))
MEMORY_POLARITIES = frozenset(("negative", "positive"))
MEMORY_SCOPES = frozenset(("contextual", "global"))
# platform_event is written by the service alone and is rejected from this client.
MEMORY_SOURCES = frozenset(("model_inferred", "user_explicit"))
MEMORY_STATUSES = frozenset(("active", "deleted", "pending_review", "rejected"))
MAX_INFERRED_CONFIDENCE = 0.6
DEFAULT_LIST_LIMIT = 20
MAX_LIST_LIMIT = 100


def _choice(data, field, allowed):
    value = text(data.get(field)).lower()
    if value not in allowed:
        raise ActionState("input_required", "%s has an unsupported value." % field, fields=[field])
    return value


def _confidence(value, source_type):
    if value in (None, ""):
        return MAX_INFERRED_CONFIDENCE if source_type == "model_inferred" else 1
    try:
        confidence = float(value)
    except (TypeError, ValueError) as error:
        raise ActionState("input_required", "confidence must be a number.", fields=["confidence"]) from error
    invalid = not math.isfinite(confidence) or confidence <= 0 or confidence > 1
    inferred_high = source_type == "model_inferred" and confidence > MAX_INFERRED_CONFIDENCE
    if invalid or inferred_high:
        raise ActionState(
            "input_required",
            "confidence must be above 0 and no more than 0.6 for inferred memory.",
            fields=["confidence"],
        )
    return confidence


def _owned_agent(api):
    agents = api.request("GET", "/agent-network-api/agents?mine=1")
    if len(agents) > 1:
        raise ActionState("failed", "The account has more than one Agent profile.", code="AGENT_ACCOUNT_INVARIANT", retryable=False)
    if not agents:
        raise ActionState("failed", "Create the account Agent profile first.", code="AGENT_PROFILE_REQUIRED", retryable=False)
    return agents[0]


def _agent_path(agent_id, suffix):
    return "/agent-network-api/agents/%s/%s" % (quote(str(agent_id), safe=""), suffix)


def _optional_text(data, field):
    return text(data.get(field)) or None


def _memory_body(data):
    source_type = _choice(data, "sourceType", MEMORY_SOURCES)
    scope = _choice(data, "scope", MEMORY_SCOPES) if data.get("scope") else "contextual"
    context_text = "" if scope == "global" else require_text(data.get("contextText"), "contextText")
    return {
        "clientMemoryId": require_text(data.get("clientMemoryId"), "clientMemoryId"),
        "confidence": _confidence(data.get("confidence"), source_type),
        "contextText": context_text,
        "conversationId": _optional_text(data, "conversationId"),
        "expiresAt": _optional_text(data, "expiresAt"),
        "kind": _choice(data, "kind", MEMORY_KINDS),
        "polarity": _choice(data, "polarity", MEMORY_POLARITIES),
        "requirementId": _optional_text(data, "requirementId"),
        "scope": scope,
        "sourceType": source_type,
        "statement": require_text(data.get("statement"), "statement"),
        "targetAgentId": _optional_text(data, "targetAgentId"),
    }


def _integer(value, field, minimum, maximum=None):
    if isinstance(value, bool):
        raise ActionState("input_required", "%s must be an integer." % field, fields=[field])
    try:
        number = float(value)
    except (TypeError, ValueError) as error:
        raise ActionState("input_required", "%s must be an integer." % field, fields=[field]) from error
    if not number.is_integer() or number < minimum or (maximum is not None and number > maximum):
        raise ActionState("input_required", "%s must be an integer in range." % field, fields=[field])
    return int(number)


def _list_query(data):
    limit = _integer(data.get("limit", DEFAULT_LIST_LIMIT), "limit", 1, MAX_LIST_LIMIT)
    params = {"limit": limit}
    if data.get("offset") is not None:
        params["offset"] = _integer(data.get("offset"), "offset", 0)
    if data.get("status"):
        params["status"] = _choice(data, "status", MEMORY_STATUSES)
    return urlencode(params)


def get_memory_settings(api, _data, _key):
    agent = _owned_agent(api)
    return api.request("GET", _agent_path(agent["id"], "memory-settings"))


def propose_memory(api, data, key):
    body = _memory_body(data)
    agent = _owned_agent(api)
    settings_path = _agent_path(agent["id"], "memory-settings")
    settings = api.request("GET", settings_path)
    if settings.get("collectionEnabled") is not True:
        raise ActionState("failed", "Memory collection is disabled for this Agent.", code="MEMORY_COLLECTION_DISABLED", retryable=False)
    return api.request(
        "POST", _agent_path(agent["id"], "memories"), body,
        key=key + ":memory:propose",
    )


def list_memories(api, data, _key):
    query = _list_query(data)
    agent = _owned_agent(api)
    path = _agent_path(agent["id"], "memories") + "?" + query
    return api.request("GET", path)


def delete_memory(api, data, key):
    memory_id = require_text(data.get("memoryId"), "memoryId")
    agent = _owned_agent(api)
    path = _agent_path(agent["id"], "memories") + "/" + quote(memory_id, safe="")
    return api.request("DELETE", path, key=key + ":memory:delete")

File: skills/tokensmind-agent-network/scripts/lib/memory-actions.js
import { actionFailure, requireInput } from './action-errors.js';
import { getMyAgent } from './agent-actions.js';
import { isObject, optionalLimit, requiredId, requiredText, text } from './action-validation.js';

const MEMORY_KINDS = new Set(['agent_experience', 'matching_preference']);
const MEMORY_POLARITIES = new Set(['negative', 'positive']);
const MEMORY_SCOPES = new Set(['contextual', 'global']);
// platform_event 由服务端独占写入，客户端提交会被拒绝。
const MEMORY_SOURCES = new Set(['model_inferred', 'user_explicit']);
const MEMORY_STATUSES = new Set(['active', 'deleted', 'pending_review', 'rejected']);
const MAX_INFERRED_CONFIDENCE = 0.6;

function requireChoice(input, field, allowed) {
  const value = text(input?.[field]).toLowerCase();
  if (!allowed.has(value)) requireInput([field], `${field} has an unsupported value.`);
  return value;
}

function optionalConfidence(value, sourceType) {
  if (value === undefined || value === null || value === '') {
    return sourceType === 'model_inferred' ? MAX_INFERRED_CONFIDENCE : 1;
  }
  const confidence = Number(value);
  const inferredLimitExceeded = sourceType === 'model_inferred'
    && confidence > MAX_INFERRED_CONFIDENCE;
  if (!Number.isFinite(confidence) || confidence <= 0 || confidence > 1 || inferredLimitExceeded) {
    requireInput(['confidence'], 'confidence must be above 0 and no more than 0.6 for inferred memory.');
  }
  return confidence;
}

async function ownedAgent(context) {
  const agent = await getMyAgent(context);
  if (!agent) actionFailure('AGENT_PROFILE_REQUIRED', 'Create the account Agent profile first.');
  return agent;
}

function agentPath(agentId, suffix) {
  return `/agent-network-api/agents/${encodeURIComponent(agentId)}/${suffix}`;
}

function optionalText(input, field) {
  return Object.hasOwn(input, field) ? text(input[field]) || null : null;
}

function memoryBody(input) {
  if (!isObject(input)) requireInput(['input']);
  const sourceType = requireChoice(input, 'sourceType', MEMORY_SOURCES);
  const scope = input.scope ? requireChoice(input, 'scope', MEMORY_SCOPES) : 'contextual';
  const contextText = scope === 'global' ? '' : requiredText(input, 'contextText');
  return {
    clientMemoryId: requiredText(input, 'clientMemoryId'),
    confidence: optionalConfidence(input.confidence, sourceType),
    contextText,
    conversationId: optionalText(input, 'conversationId'),
    expiresAt: optionalText(input, 'expiresAt'),
    kind: requireChoice(input, 'kind', MEMORY_KINDS),
    polarity: requireChoice(input, 'polarity', MEMORY_POLARITIES),
    requirementId: optionalText(input, 'requirementId'),
    scope,
    sourceType,
    statement: requiredText(input, 'statement'),
    targetAgentId: optionalText(input, 'targetAgentId'),
  };
}

function listQuery(input) {
  const params = new URLSearchParams({ limit: String(optionalLimit(input.limit)) });
  if (input.offset !== undefined) {
    const offset = Number(input.offset);
    if (!Number.isInteger(offset) || offset < 0) requireInput(['offset']);
    params.set('offset', String(offset));
  }
  if (input.status) params.set('status', requireChoice(input, 'status', MEMORY_STATUSES));
  return params.toString();
}

export async function getMemorySettings(context) {
  const agent = await ownedAgent(context);
  return context.get(agentPath(agent.id, 'memory-settings'));
}

export async function proposeMemory(context) {
  const body = memoryBody(context.input);
  const agent = await ownedAgent(context);
  const settings = await context.get(agentPath(agent.id, 'memory-settings'));
  if (settings?.collectionEnabled !== true) {
    actionFailure('MEMORY_COLLECTION_DISABLED', 'Memory collection is disabled for this Agent.');
  }
  return context.mutate({
    method: 'POST',
    path: agentPath(agent.id, 'memories'),
    body,
    label: 'memory:propose',
  });
}

export async function listMemories(context) {
  const query = listQuery(context.input);
  const agent = await ownedAgent(context);
  return context.get(`${agentPath(agent.id, 'memories')}?${query}`);
}

export async function deleteMemory(context) {
  const memoryId = requiredId(context.input, 'memoryId');
  const agent = await ownedAgent(context);
  return context.mutate({
    method: 'DELETE',
    path: `${agentPath(agent.id, 'memories')}/${encodeURIComponent(memoryId)}`,
    label: 'memory:delete',
  });
}
